README.md
# dicom-to-json

The nodejs dcm2json native modules of dcmtk.

# Complier environment
- nodejs v14.15.0

## Windows
- Windows 10 64bits (vs 2019)

## Linux
- Ubuntu 20.04 (g++ 9.3.0)

### Install iconv
Before using, you must install the iconv
- download iconv
```sh
wget https://ftp.gnu.org/pub/gnu/libiconv/libiconv-1.16.tar.gz
```
- tar gz
```sh
tar zxvf libiconv-1.16.tar.gz
```
- config
```
cd libiconv-1.16 \
./configure --prefix=/usr/local/apps
```
- complier
```sh
sudo make && sudo make install
```

# Troubleshooting
## F: Unknown VR: Tag not found in data dictionary: converting file to UTF-8:
- The DCMDICTPATH environment variable is not set.
- You can run command below
```sh
export DCMDICTPATH=$(pwd)/dmcdata/data/dicom.dic
```

# example
## dcm2json
```js
const { dcm2json } = require("dicom-to-json");
const path=  require('path');

dcm2json(path.join(__dirname, "/000001.dcm") , function (json) {
    console.log(json);
})
```
### output
<details>
    <summary>click to show</summary>

  ```json
  I: M:\nodejs\dicom-module-converter\example\0003.dcm
W: DcmSequenceOfItems: Length of item in sequence PixelData (7fe0,0010) is odd
W: DcmSequenceOfItems: Length of item in sequence PixelData (7fe0,0010) is odd
W: DcmSequenceOfItems: Length of item in sequence PixelData (7fe0,0010) is odd
W: DcmSequenceOfItems: Length of item in sequence PixelData (7fe0,0010) is odd
W: DcmSequenceOfItems: Length of item in sequence PixelData (7fe0,0010) is odd
W: DcmSequenceOfItems: Length of item in sequence PixelData (7fe0,0010) is odd
W: DcmSequenceOfItems: Length of item in sequence PixelData (7fe0,0010) is odd
W: DcmSequenceOfItems: Length of item in sequence PixelData (7fe0,0010) is odd
W: DcmSequenceOfItems: Length of item in sequence PixelData (7fe0,0010) is odd
W: DcmSequenceOfItems: Length of item in sequence PixelData (7fe0,0010) is odd
I: converting all element values that are affected by SpecificCharacterSet (0008,0005) to UTF-8
I: get json success
{
  "00020001": {
    "vr": "OB",
    "InlineBinary": "AAE="
  },
  "00020002": {
    "vr": "UI",
    "Value": [
      "1.2.840.10008.5.1.4.1.1.12.1"
    ]
  },
  "00020003": {
    "vr": "UI",
    "Value": [
      "1.3.12.2.1107.5.4.3.284980.19951129.170916.9"
    ]
  },
  "00020010": {
    "vr": "UI",
    "Value": [
      "1.2.840.10008.1.2.4.50"
    ]
  },
  "00020012": {
    "vr": "UI",
    "Value": [
      "1.3.12.2.1107.5.4.3.2"
    ]
  },
  "00020016": {
    "vr": "AE",
    "Value": [
      "SIEMENS:DCR 1.0"
    ]
  },
  "00080005": {
    "vr": "CS",
    "Value": [
      "ISO_IR 192"
    ]
  },
  "00080008": {
    "vr": "CS",
    "Value": [
      "DERIVED",
      "PRIMARY",
      "BIPLANE A"
    ]
  },
  "00080016": {
    "vr": "UI",
    "Value": [
      "1.2.840.10008.5.1.4.1.1.12.1"
    ]
  },
  "00080018": {
    "vr": "UI",
    "Value": [
      "1.3.12.2.1107.5.4.3.284980.19951129.170916.9"
    ]
  },
  "00080020": {
    "vr": "DA",
    "Value": [
      "19930325"
    ]
  },
  "00080030": {
    "vr": "TM",
    "Value": [
      "135731"
    ]
  },
  "00080050": {
    "vr": "SH"
  },
  "00080060": {
    "vr": "CS",
    "Value": [
      "XA"
    ]
  },
  "00080070": {
    "vr": "LO"
  },
  "00080080": {
    "vr": "LO",
    "Value": [
      "INSTITUTION                 ."
    ]
  },
  "00080081": {
    "vr": "ST"
  },
  "00080090": {
    "vr": "PN"
  },
  "00081030": {
    "vr": "LO",
    "Value": [
      "8 images of 4 biplane acqs"
    ]
  },
  "00081050": {
    "vr": "PN",
    "Value": [
      {
        "Alphabetic": "Well^U^R^Dr."
      }
    ]
  },
  "00081140": {
    "vr": "SQ",
    "Value": [
      {
        "00081150": {
          "vr": "UI",
          "Value": [
            "1.2.840.10008.5.1.4.1.1.12.1"
          ]
        },
        "00081155": {
          "vr": "UI",
          "Value": [
            "1.3.12.2.1107.5.4.3.284980.19951129.170916.11"
          ]
        }
      }
    ]
  },
  "00082110": {
    "vr": "CS",
    "Value": [
      "01"
    ]
  },
  "00082112": {
    "vr": "SQ",
    "Value": [
      {
        "00081150": {
          "vr": "UI",
          "Value": [
            "1.2.840.10008.5.1.4.1.1.12.1"
          ]
        },
        "00081155": {
          "vr": "UI",
          "Value": [
            "1.3.12.2.1107.5.4.3.284980.19951129.170916.8"
          ]
        }
      }
    ]
  },
  "00090010": {
    "vr": "LO",
    "Value": [
      "CARDIO-SMS 1.0"
    ]
  },
  "00091002": {
    "vr": "OB",
    "InlineBinary": "AAAAAQAAAAEAAAAAAAAAAQAAAAAAAAABAAAAAAAAAAEAAAABAAAAAAAAAAAAAAADAAAA/wAAAAAAAAB/AAAA/wAAAAEAAAAAAAAAAAAAAAAAAAADAAAACAAAAAEAAAAAAAAAAAAAAAAAAAADAAAACC0tLS0tLS0tLgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADMyMi0yOTItNzMtNgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADE5NTgwNzE5AAAzNAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABGAAAAAAAAAAAAAAAAAAAAAABJTlNUSVRVVElPTgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAtLS0tLS0tLS0tLS4AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAxOTkzMDMyNQAAMTM1NzMxAAAAAAAAAAAAAAAAMTk5MzAzMjUAADEzNTczMQAAAAAAAAAAAAAAADggaW1hZ2VzIG9mIDQgYmlwbGFuZSBhY3FzAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABTTVNSQTAwMzM1AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAIAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAACAAABEQAAAAAxLjMuMTIuMi4xMTA3LjUuNC4zLjQ5NzUzMTY3NzcyMTYuMTk5NTExMTQuOTQxMDEuMTYAAAAAAAAAAAAAAAAAAAAxLjMuMTIuMi4xMTA3LjUuNC4zLjQ5NzUzMTY3NzcyMTYuMTk5NTExMTQuOTQxMDEuMTcAAAAAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=="
  },
  "00091003": {
    "vr": "OB",
    "InlineBinary": "AAAAATE5OTMwMzI1AAAxMzU3MzEAAAAAAAAAAAAAAAAxOTkzMDMyNQAAMTM1NzMxAAAAAAAAAAAAAAAASGljb3IgRENNAAAAAAAAAAAAAAAAAAAAAAAAAQAAAAAAAAABARE0WAAAAAUBETRYAR8C/AAAANgAAAAEARE6aAAAAAEAAAADAAAAAAAAAAAAAAAAAAAABQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA/wAAAAAAAAAAAAAAAAAAABEAAAABAAAAQQAAAAAAAAAAAAAF3AAAAAEAAAAAAAAAAAAAAAAAAAABAAAAAQAAASr///zbAAABHQAAAAMAAADQAAAA0QAAA4gAAAOZAAAEFQAAAAAAAAAAAAAGFgAACrwAAAAtAAAAAAAAAAAAAAAAAAAAAQAAAAEAAAAAAAAAAQAAAAcAAAAAAAAAAAAAAAAAAAABAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACIAAAACAAAABAAAAAAAAAD/AAAAAAAAAAQAAAAAAAAA/wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAPAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAAAAAAAAABARE1/AAAAAUBETX8"
  },
  "00091005": {
    "vr": "OB",
    "InlineBinary": "AAAFUAAAABEAAAAEAAAEsAAAAFAAAAAC"
  },
  "00100010": {
    "vr": "PN",
    "Value": [
      {
        "Alphabetic": "Rubo DEMO"
      }
    ]
  },
  "00100020": {
    "vr": "LO",
    "Value": [
      "322-292-73-6"
    ]
  },
  "00100030": {
    "vr": "DA",
    "Value": [
      "19580719"
    ]
  },
  "00100040": {
    "vr": "CS",
    "Value": [
      "F"
    ]
  },
  "00180060": {
    "vr": "DS"
  },
  "00181063": {
    "vr": "DS",
    "Value": [
      66
    ]
  },
  "00181152": {
    "vr": "IS"
  },
  "00181155": {
    "vr": "CS",
    "Value": [
      "GR"
    ]
  },
  "00181500": {
    "vr": "CS"
  },
  "00181510": {
    "vr": "DS",
    "Value": [
      29.8
    ]
  },
  "00181511": {
    "vr": "DS",
    "Value": [
      28.5
    ]
  },
  "00190010": {
    "vr": "LO",
    "Value": [
      "CARDIO-D.R. 1.0"
    ]
  },
  "00191030": {
    "vr": "UL",
    "Value": [
      262144
    ]
  },
  "0020000D": {
    "vr": "UI",
    "Value": [
      "1.3.12.2.1107.5.4.3.4975316777216.19951114.94101.16"
    ]
  },
  "0020000E": {
    "vr": "UI",
    "Value": [
      "1.3.12.2.1107.5.4.3.4975316777216.19951114.94101.17"
    ]
  },
  "00200010": {
    "vr": "SH"
  },
  "00200011": {
    "vr": "IS",
    "Value": [
      1
    ]
  },
  "00200013": {
    "vr": "IS"
  },
  "00200020": {
    "vr": "CS"
  },
  "00210010": {
    "vr": "LO",
    "Value": [
      "CARDIO-D.R. 1.0"
    ]
  },
  "00211013": {
    "vr": "IS",
    "Value": [
      5
    ]
  },
  "00280002": {
    "vr": "US",
    "Value": [
      1
    ]
  },
  "00280004": {
    "vr": "CS",
    "Value": [
      "MONOCHROME2"
    ]
  },
  "00280008": {
    "vr": "IS",
    "Value": [
      17
    ]
  },
  "00280009": {
    "vr": "AT",
    "Value": [
      "00181063"
    ]
  },
  "00280010": {
    "vr": "US",
    "Value": [
      512
    ]
  },
  "00280011": {
    "vr": "US",
    "Value": [
      512
    ]
  },
  "00280100": {
    "vr": "US",
    "Value": [
      8
    ]
  },
  "00280101": {
    "vr": "US",
    "Value": [
      8
    ]
  },
  "00280102": {
    "vr": "US",
    "Value": [
      7
    ]
  },
  "00280103": {
    "vr": "US",
    "Value": [
      0
    ]
  },
  "00281040": {
    "vr": "CS",
    "Value": [
      "LIN"
    ]
  },
  "00281090": {
    "vr": "CS",
    "Value": [
      "NAT"
    ]
  },
  "00286040": {
    "vr": "US",
    "Value": [
      1,
      12
    ]
  },
  "00286100": {
    "vr": "SQ",
    "Value": [
      {
        "00286101": {
          "vr": "CS",
          "Value": [
            "NONE"
          ]
        },
        "00286110": {
          "vr": "US",
          "Value": [
            0
          ]
        }
      }
    ]
  },
  "00290010": {
    "vr": "LO",
    "Value": [
      "CARDIO-D.R. 1.0"
    ]
  },
  "00291000": {
    "vr": "SQ",
    "Value": [
      {
        "00290010": {
          "vr": "LO",
          "Value": [
            "CARDIO-D.R. 1.0"
          ]
        },
        "00291001": {
          "vr": "US",
          "Value": [
            5,
            5
          ]
        },
        "00291002": {
          "vr": "US",
          "Value": [
            1,
            1,
            1,
            1,
            1,
            1,
            1,
            1,
            1,
            1,
            1,
            1,
            1,
            1,
            1,
            1,
            1,
            1,
            1,
            1,
            1,
            1,
            1,
            1,
            1
          ]
        },
        "00291003": {
          "vr": "FL",
          "Value": [
            0.5
          ]
        }
      }
    ]
  },
  "50000005": {
    "vr": "US",
    "Value": [
      2
    ]
  },
  "50000010": {
    "vr": "US",
    "Value": [
      1360
    ]
  },
  "50000020": {
    "vr": "CS",
    "Value": [
      "ECG"
    ]
  },
  "50000030": {
    "vr": "SH",
    "Value": [
      "DPPS",
      "NONE"
    ]
  },
  "50000103": {
    "vr": "US",
    "Value": [
      0
    ]
  },
  "50000104": {
    "vr": "US"
  },
  "50000105": {
    "vr": "US"
  },
  "50000106": {
    "vr": "US"
  },
  "50000110": {
    "vr": "US",
    "Value": [
      0,
      1
    ]
  },
  "50000112": {
    "vr": "US",
    "Value": [
      0
    ]
  },
  "50000114": {
    "vr": "US",
    "Value": [
      80
    ]
  },
  "50003000": {
    "vr": "OW",
    "InlineBinary": "/wD/AP8A/wD/AP8A/wD6AOoA2wDJALgApgCQAH4AbQBgAFUATABDADgALgAmACAAGwAXABQAEgARABIAEQAQABAADwAPABIAEQAQAA8AEAARABIAFAAVABQAEgAUABUAFgAYABkAGwAdAB0AHgAfAB8AHgAgACEAIgAiACMAIgAgAB4AHgAdABwAGgAaABkAFwAUABIADwALAAsADgAPABAAEQATABYAGAAYABoAGgAbABoAGAAYABYAEwASABIAEQARABAAEAAQABEAEwAWABYAFQATABEAEAAQAA8ADgANAAsACgALABAAEwAUABQAEwARABAAEQARAA8ADAAKAAsADAAQABIAFQAXABkAGAAXABYAFgAVABUAFAATABIAEQAQABAAEQASABMAEwAUABQAFgAXABgAGQAbABwAHAAbAB4AIgAlACkAKwAtAC4AMAAzADQANwA5ADkAOgA8AD0AQABDAEMAQgA/AD0APQA9ADsAOwA5ADYANgA/AEgAUABXAFwAXwBfAGAAYABfAF0AWgBZAFgAVgBSAE4ASgBHAEQAQgBBAEIAQwBHAEwATgBOAE4ASgBHAEQAQwBCAEAAPAA4ADYANAAzADIANAA2ADUAMwA0ADUANQA4ADsAPgA/AEAAQwBFAEYASgBLAEsATABPAFMAWQBeAGIAZABlAGcAaABqAG0AbwBwAHEAcwB0AHYAdgB2AHgAeQB6AHsAewB5AHgAdgB0AHIAcABtAG0AbgBvAG8AbgBuAGwAbABuAG8AcQBxAHEAcQByAHEAcABvAG4AawBoAGcAYwBfAFwAWwBZAFYAUwBNAEgAQwBBAEAAQAA+ADsAOQA2ADMAMAAvAC8ALgAsACwALAAsACsAKgAqACoAKgAoACgAJwAmACcAKAAoACYAIwAiACAAHgAdABsAGAAUABEAEgARABEAEgASAA8ADQAMAAoACQAHAAUABAAFAAMAAwAEAAUABwAHAAYABwAJAAkACwANAA4ADgANAAwACwALAAsACwAMAA0AEgAXABwAIAAhAB8AHQAaABkAFwAYABgAFgAVABQAEwASABEADwANAAsACQAIAAYAAwABAAEAAQACAAIAAgACAAIAAgABAAAAAAAAAAAAAAABAAIAAwABAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQABAAEAAAAAAAAAAgAFAAcABwAHAAcABgAFAAUACwAQABIAFQAZABsAGwAcABwAGgAYABYAFAARAA8ADQAOAA4ADQAMAAsACQAHAAUABAADAAEAAAABAAMABAAGAAcABwAGAAYABwAIAAoADAAOAA8ADwARABMAFgAXABgAFwAWABcAGAAYABgAFwAVABMAEAANAAoACAAFAAMAAAAAAAAAAAABAAMABQAIAAoACwAMAAsACwAMAA8AEgAUABUAFAATABMAFAAUABMAFAATABAAEAATABUAFQAUABMAEQAPAA8ADwAOAA0ADQAPABAAEQATABQAEgAQABAADwAQABMAFQAZAB0AHwAfACAAIAAgACAAIAAhACAAHgAgACEAIQAgAB4AGwAXABYAFgAWABYAFQAWABgAGAAXABYAGAAZABkAGgAYABQAEQAOAA0ACwAJAAcABgAEAAUABgAHAAgACQAHAAYACAAKAAsADAAMAAsACgAKAA0ADgAOAA8ADwAOAA4AEQATABYAGAAZABUAEgAPAAwADAAPABEAFAAWABkAGwAfACMAJgArAC0AMAAzADYANwA4ADkAOgA8AD4APwBBAEIARQBKAE8AUQBSAFMAVABVAFYAVgBVAFQAVABUAFYAWQBaAFgAVQBTAFQAVABXAFwAXABbAFwAXwBgAGEAYQBhAGAAXgBcAFsAWABVAFQAVABTAE8ASwBLAEsASwBKAEgARwBEAEMAQQA9ADsAOQA2ADAAKwAnACQAIgAgAB8AHQAbABsAGgAZABcAFQATABIADwALAAkACAAIAAkACQAIAAUAAwAHAAsADQANAAwACQAHAAgACQAJAAgABQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABAAQABgAGAAcACAAJAAwADwAQAA8ADgAOAA4ADgANAAwABwAEAAEAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEABAAHAApADkARgBXAGcAdQCCAI8AnQCoALcAxADSAOAA6wD1AP4A/wD/AP8A/QD2AOkA3ADQAMcAvQCwAKMAkgB+AG4AXwBSAEYAPQA2ADAAKwAiABgADwAGAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABgAKAAwADQAMAAsADAANAA4ADQAJAAcABgAGAAcACQAJAAgACAAHAAcABwAHAAcABwAHAAYABQAGAAgACAAKAAsACgAJAAgACQAKAAoACgALAAkACAALAA4AEAATABQAEgAQAA4ADAAMAA8AEAARABMAEwASABAAEgASABMAFAAVABcAFwAbACAAIwAkACMAHgAYABYAEwAQAA4ADAAIAAcABwAIAAgABgAEAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQACAAEAAQADAAQABgAGAAcACAAJAAYABAAEAAUABQAFAAQABAAFAAUABwAIAAsADQANAAwACwANAA8AEgAVABYAFgAVABcAGQAaABoAGQAWABQAFAAVABgAGwAeAB4AHgAfACAAHwAjACcAKAApACsAKwArAC0ALwAxADMANAA0ADUANwA6AD4AQABCAEIAQgBDAEQARQBGAEcARgBEAEUASQBMAEwATABJAEIAPgA9AD4AQABEAEcASQBKAE0ATgBQAFAAUABQAFAAUABPAE4ATwBPAE0ATABLAEoARwBHAEUAQwBBAD4APAA7ADsAPAA+AD8AQABCAEIARABFAEcASgBOAE8AUQBRAFIAUgBTAFMAUgBQAE8AUABRAFQAVQBVAFUAVABTAFMAVABUAFQAVABUAFYAVwBZAFoAWwBaAFoAVgBSAE8ATwBPAFIAVABTAE4ASQBEAEAAPgA9ADsANgAyAC8ALQAtACwAKgAoACQAHwAaABcAFQAUABAADwAOAA0ADQANAA0ADAAKAAcABQABAAEAAgAEAAUABQADAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABAAHAAgACQAHAAUAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="
  }
}
  ```
</details>

## dcm2xml
```js
const { dcm2xml } = require("../index");
const path=  require('path');

dcm2xml(path.join(__dirname, "/0003.dcm") , function (json) {
    console.log(json);
})
```

### output
<details>
    <summary>click to show</summary>

    ```xml
    I: M:\nodejs\dicom-module-converter\example\0003.dcm
    W: DcmSequenceOfItems: Length of item in sequence PixelData (7fe0,0010) is odd
    W: DcmSequenceOfItems: Length of item in sequence PixelData (7fe0,0010) is odd
    W: DcmSequenceOfItems: Length of item in sequence PixelData (7fe0,0010) is odd
    W: DcmSequenceOfItems: Length of item in sequence PixelData (7fe0,0010) is odd
    W: DcmSequenceOfItems: Length of item in sequence PixelData (7fe0,0010) is odd
    W: DcmSequenceOfItems: Length of item in sequence PixelData (7fe0,0010) is odd
    W: DcmSequenceOfItems: Length of item in sequence PixelData (7fe0,0010) is odd
    W: DcmSequenceOfItems: Length of item in sequence PixelData (7fe0,0010) is odd
    W: DcmSequenceOfItems: Length of item in sequence PixelData (7fe0,0010) is odd
    W: DcmSequenceOfItems: Length of item in sequence PixelData (7fe0,0010) is odd
    I: converting all element values that are affected by SpecificCharacterSet (0008,0005) to UTF-8
    I: get xml success
    <?xml version="1.0" encoding="UTF-8"?>
    <NativeDicomModel xml:space="preserve">
    <DicomAttribute tag="00080005" vr="CS" keyword="SpecificCharacterSet">
    <Value number="1">ISO_IR 192</Value>
    </DicomAttribute>
    <DicomAttribute tag="00080008" vr="CS" keyword="ImageType">
    <Value number="1">DERIVED</Value>
    <Value number="2">PRIMARY</Value>
    <Value number="3">BIPLANE A</Value>
    </DicomAttribute>
    <DicomAttribute tag="00080016" vr="UI" keyword="SOPClassUID">
    <Value number="1">1.2.840.10008.5.1.4.1.1.12.1</Value>
    </DicomAttribute>
    <DicomAttribute tag="00080018" vr="UI" keyword="SOPInstanceUID">
    <Value number="1">1.3.12.2.1107.5.4.3.284980.19951129.170916.9</Value>
    </DicomAttribute>
    <DicomAttribute tag="00080020" vr="DA" keyword="StudyDate">
    <Value number="1">19930325</Value>
    </DicomAttribute>
    <DicomAttribute tag="00080030" vr="TM" keyword="StudyTime">
    <Value number="1">135731</Value>
    </DicomAttribute>
    <DicomAttribute tag="00080050" vr="SH" keyword="AccessionNumber">
    </DicomAttribute>
    <DicomAttribute tag="00080060" vr="CS" keyword="Modality">
    <Value number="1">XA</Value>
    </DicomAttribute>
    <DicomAttribute tag="00080070" vr="LO" keyword="Manufacturer">
    </DicomAttribute>
    <DicomAttribute tag="00080080" vr="LO" keyword="InstitutionName">
    <Value number="1">INSTITUTION                 .</Value>
    </DicomAttribute>
    <DicomAttribute tag="00080081" vr="ST" keyword="InstitutionAddress">
    </DicomAttribute>
    <DicomAttribute tag="00080090" vr="PN" keyword="ReferringPhysicianName">
    </DicomAttribute>
    <DicomAttribute tag="00081030" vr="LO" keyword="StudyDescription">
    <Value number="1">8 images of 4 biplane acqs</Value>
    </DicomAttribute>

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.2.0](https://github.com/Chinlinlee/dicom-to-json/compare/v1.1.0...v1.2.0) (2022-04-09)


### Features

* add dcm2xml, refactor napi export item ([f336c76](https://github.com/Chinlinlee/dicom-to-json/commit/f336c76841c62b15ba6af0246cbd6a79106f8009))
    * use class to export dcm2json and dcm2xml


src/dcm2json.cc
/*
*
*  Copyright (C) 2016-2020, OFFIS e.V.
*  All rights reserved.  See COPYRIGHT file for details.
*
*  This software and supporting documentation were developed by
*
*    OFFIS e.V.
*    R&D Division Health
*    Escherweg 2
*    D-26121 Oldenburg, Germany
*
*
*  Module:  dcmdata
*
*  Author:  Sebastian Grallert
*
*  Purpose: Convert the contents of a DICOM file to JSON format
*
*/

#include "dcmtk/config/osconfig.h" /* make sure OS specific configuration is included first */

#include "dcmtk/dcmdata/dctk.h"
#include "dcmtk/dcmdata/cmdlnarg.h"

#include "dcmtk/dcmdata/dcjson.h"
#include "dcmtk/ofstd/ofstream.h"
#include "dcmtk/ofstd/ofconapp.h"
#include "dcmtk/ofstd/ofexit.h"

#ifdef WITH_ZLIB
#include <zlib.h> /* for zlibVersion() */
#endif
#ifdef DCMTK_ENABLE_CHARSET_CONVERSION
#include "dcmtk/ofstd/ofchrenc.h" /* for OFCharacterEncoding */
#endif

#define OFFIS_CONSOLE_APPLICATION "dcm2json"
#define OFFIS_CONSOLE_DESCRIPTION "Convert DICOM file and data set to JSON"

#define EXITCODE_CANNOT_CONVERT_TO_UNICODE 80
#define EXITCODE_CANNOT_WRITE_VALID_JSON 81

static OFLogger dcm2jsonLogger = OFLog::getLogger("dcmtk.apps." OFFIS_CONSOLE_APPLICATION);

// ********************************************

/* Function to call all writeJson() functions in DCMTK */
static OFCondition writeFile(STD_NAMESPACE ostream &out,
                             const char *ifname,
                             DcmFileFormat *dfile,
                             const E_FileReadMode readMode,
                             const OFBool format,
                             const OFBool printMetaInfo,
                             const OFBool encode_extended)
{
    
    OFCondition result = EC_IllegalParameter;
    if ((ifname != NULL) && (dfile != NULL))
    {
        /* write JSON document content */
        DcmDataset *dset = dfile->getDataset();
        if (format)
        {
            DcmJsonFormatPretty fmt(printMetaInfo);
            fmt.setJsonExtensionEnabled(encode_extended);
            if (readMode == ERM_dataset)
                result = dset->writeJsonExt(out, fmt, OFTrue, OFTrue);
            else
                result = dfile->writeJson(out, fmt );
        }
        else
        {
            DcmJsonFormatCompact fmt(printMetaInfo);
            fmt.setJsonExtensionEnabled(encode_extended);
            if (readMode == ERM_dataset)
                result = dset->writeJsonExt(out, fmt, OFTrue, OFTrue );
            else
                result = dfile->writeJson(out, fmt );
        }
    }
    return result;
}

#define SHORTCOL 3
#define LONGCOL 20

static char rcsid[] = "$dcmtk: " OFFIS_CONSOLE_APPLICATION " v" OFFIS_DCMTK_VERSION " " OFFIS_DCMTK_RELEASEDATE " $";

#include <assert.h>
#include <node_api.h>
#include "dcm2json.h"


char *getDCMJson(char *ifname)
{
    OFBool opt_format = OFTrue;
    OFBool opt_addMetaInformation = OFTrue;
    OFBool opt_encode_extended = OFFalse;
    E_FileReadMode opt_readMode = ERM_autoDetect;
    E_TransferSyntax opt_ixfer = EXS_Unknown;
    OFString optStr;

    OFConsoleApplication app(OFFIS_CONSOLE_APPLICATION, OFFIS_CONSOLE_DESCRIPTION, rcsid);
    int result = 0;
    if ((ifname != NULL) && (strlen(ifname) > 0))
    {
        DcmFileFormat dfile;
        OFCondition status = dfile.loadFile(ifname, opt_ixfer, EGL_noChange, DCM_MaxReadLength, opt_readMode);
        if (status.good())
        {
            DcmDataset *dset = dfile.getDataset();
            OFString csetString;
            if (dset->findAndGetOFStringArray(DCM_SpecificCharacterSet, csetString).good())
            {
                if (csetString.compare("ISO_IR 6") == 0) 
                {
                    if (dset->containsExtendedCharacters(OFFalse /*checkAllStrings*/))
                    {
                        OFLOG_FATAL(dcm2jsonLogger, "dataset contains extended characters but SpecificCharacterSet (0008,0005) is 'ISO_IR 6'");
                        result = EXITCODE_CANNOT_CONVERT_TO_UNICODE;
                        return "80";
                    }
                }
                else if (csetString.compare("ISO_IR 192") == 0)
                {

                }
                else 
                {
                                        /* we have a character set other than ASCII or UTF-8. Perform conversion. */
#ifdef DCMTK_ENABLE_CHARSET_CONVERSION
                    /* convert all DICOM strings to UTF-8 */
                    OFLOG_INFO(dcm2jsonLogger, "converting all element values that are affected by SpecificCharacterSet (0008,0005) to UTF-8");
                    status = dset->convertToUTF8();
                    if (status.bad())
                    {
                        OFLOG_FATAL(dcm2jsonLogger, status.text() << ": converting file to UTF-8: " << ifname);
                        result = EXITCODE_CANNOT_CONVERT_TO_UNICODE;
                        return "80";
                    }
#else
                    OFLOG_FATAL(dcm2jsonLogger, "character set conversion not available");
                    result = EXITCODE_CANNOT_CONVERT_TO_UNICODE;
#endif
                }
            }
            if (result == 0)
            {
                std::ostringstream stream;
                status = writeFile(stream, ifname, &dfile, opt_readMode, opt_format, opt_addMetaInformation, opt_encode_extended);
                if (status.bad())
                {
                    OFLOG_FATAL(dcm2jsonLogger, status.text() << ": " << ifname);
                    result = EXITCODE_CANNOT_WRITE_VALID_JSON;
                    return "81";
                }
                //COUT << stream.str();
                auto  streamStr = stream.str();
                char *item = new char[streamStr.size() +1 ];
                //important!
                //use strcpy avoid char* empty
                strcpy(item , streamStr.c_str());
                OFLOG_INFO(dcm2jsonLogger , "get json success");
                return item;
            }
        }
        else
        {
            OFLOG_FATAL(dcm2jsonLogger, status.text() << ": " << ifname);
            return "2";
        }
    }
    return NULL;
}



napi_value DCM2JSON::dcm2json(napi_env env, napi_callback_info info)
{
    napi_status napiStatus;
    napi_value jsonResult[1];
    size_t argc = 2;
    napi_value args[2];
    napiStatus = napi_get_cb_info(env, info, &argc, args, NULL, NULL);
    char *ifname = (char *)malloc(741478763* sizeof(char));
    size_t str_size;
    napi_value cb = args[1];
    napiStatus = napi_get_value_string_utf8(env, args[0], ifname, 741478763, &str_size);
    OFLOG_INFO(dcm2jsonLogger, ifname);
    char* dcmjson = getDCMJson(ifname);
    if (dcmjson == "81") 
    {
        napiStatus = napi_throw_error(env, "81" , "EXITCODE_CANNOT_WRITE_VALID_JSON");
    }
    else if (dcmjson == "80") 
    {
        napiStatus = napi_throw_error(env, "81" , "EXITCODE_CANNOT_CONVERT_TO_UNICODE");
    }
    else if (dcmjson == "2") 
    {
        napiStatus = napi_throw_error(env, "404" , "No such file or directory or not a valid DICOM file");
    }
    else
    {
        napiStatus = napi_create_string_utf8(env, dcmjson, NAPI_AUTO_LENGTH, jsonResult);
        napi_value global;
        napiStatus = napi_get_global(env, &global);
        napi_value cbresult;
        napiStatus = napi_call_function(env, global, cb, 1, jsonResult, &cbresult);
    }
    return NULL;
}